src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs

```diff
@@ -159,48 +118,43 @@ public override ImmutableArray<ParameterSymbol> Parameters
             get
             {
                 ComputeParameters();
-                return _lazyParametersAndDiagnostics.Parameters;
+                return _lazyParameters;
             }
         }
 
         private void ComputeParameters()
         {
-            if (_lazyParametersAndDiagnostics != null)
+            lock (_declarationDiagnostics)
```

Comment:
>lock (_declarationDiagnostics) [](start = 12, length = 30)

To avoid possibility of getting into a dead-lock situations it is probably better to calculate everything into locals, including collecting diagnostics into a local bag, and then only store the result while the lock is taken.